content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -391,24 +400,209 @@ Here's the generic procedure again, now showing the overlay:
+storage element to a stand-alone Boolean variable. The other seven bits
+are simply not used by that variable. Moreover, those seven bits are not
+used by any other stand-alone object either, because access would be far
+less efficient, and such sharing would require some kind of locking to
+prevent tasks from interfering with each other when accessing those
+stand-alone objects. (Stand-alone objects are independently addressable;
+they wouldn't stand alone otherwise.)
+
+By the same token (and still assuming a 32-bit machine), a compiler will
+allocate more than 31 bits to a variable of subtype Natural because
+there is no 31-bit addressable unit. The variable will get all 32-bits.
+
+Note that we're talking about individual, stand-alone variables.
+Components of composite types, on the other hand, might indeed share
+bytes if the individual components don't require all the bits of their
+storage elements. You'd have to request that representation, though,
+with most implementations, because accessing the components at run-time
+would require more machine instructions. We'll go into the details of
+that later.
+
+Let's talk further about types.
+
+For record types, :ada:`'Size` gives the minimum number of bits required
+to represent the whole composite value. But again, that's not
+necessarily the number of bits required for the objects' in-memory
+representation. The order of the components within the record can make
+a difference, as well as their alignments. The compiler will respect the
+alignment requirements of the components, and may add padding bytes
+within the record and also at the end to ensure components start at
+addresses compatible with their alignment requirements. As a result the
+overall size could be larger.
+
+Note that Ada compilers are allowed to reorder the components; the order
+in memory might not match the order in the source code.
+
+For example, consider this record type and its components:
+
+.. image:: images/unoptimized-record-component-order.png
+  :width: 600
+  :alt: Memory allocated to a record with unoptimized layout
+
+In the figure, we see a record type with some components, and a sample
+layout for that record type assuming the compiler does not reorder the
+components. Observe that some bytes allocated to objects of type R are
+unused (the darkly shaded ones). In this case that's because the
+alignment of subtype S happens to be 4 on this machine. The component X
+of that subtype S cannot start at byte offset 1, or 2, or 3, because
+those addresses would not satisfy the alignment constraint of S. (We're
+assuming byte 0 is at a word-aligned address.) Therefore, X starts at
+the object's starting address plus 4. Components B and C are of types
+that have an alignment of 1, so they can start at any storage element.
+They immediately follow the bytes allocated to component X. Therefore,
+R'Size is 80, or 10 bytes. The three bytes following component M are simply not
+used.
```

Comment:
```suggestion
alignment of subtype :ada:`S` happens to be 4 on this machine. The component
:ada:`X` of that subtype :ada:`S` cannot start at byte offset 1, or 2, or 3, 
because those addresses would not satisfy the alignment constraint of 
:ada:`S`. (We're assuming byte 0 is at a word-aligned address.) Therefore,
:ada:`X` starts at the object's starting address plus 4. Components
:ada:`B` and :ada:`C` are of types that have an alignment of 1, so they 
can start at any storage element.
They immediately follow the bytes allocated to component :ada:`X`. 
Therefore, :ada:`R'Size` is 80, or 10 bytes. The three bytes following 
component :ada:`M are simply not used.
```